Refactoring `BorderControl`'s unit tests

andrewhayward opened the pull request:
## What?
This patch migrates uses of `fireEvent` in `BorderControl` to [`testing-library/user-event`](https://testing-library.com/docs/user-event), where possible; there is one remaining usage around sliders/range inputs that [isn't currently supported](https://github.com/testing-library/user-event/issues/871) by `userEvent`.

## Why?
We should, where feasible, be [using `userEvent` over `fireEvent`](https://ph-fritsche.github.io/blog/post/why-userevent), as it better replicates how users interact with the DOM.

## How?
The `BorderControl` test file has been refactored to (mostly) replace `fireEvent` calls with an equivalent use of `userEvent`.

## Testing Instructions
Nothing user-facing is changed, only automated tests. This should be covered by the CI running against this PR.

Commit:
Refactoring `BorderControl`'s unit tests

This patch migrates uses of `fireEvent` in `BorderControl` to `testing-library/user-event`, where possible; there is one remaining usage that isn't currently supported by `userEvent`.

Commit:
Updating failing tests – user error

Was running the wrong tests locally

Commit:
Merge branch 'trunk' into internal/refactor-bordercontrol-unit-tests-fireevent-userevent

andrewhayward commented:
@ciampo @mirka would appreciate your feedback here – not used `userEvent` before.

ciampo commented on packages/components/src/border-control/test/index.js:
```diff
@@ -415,53 +407,62 @@ describe( 'BorderControl', () => {
+			await user.type( widthInput, '0' );
 
-			expect( props.onChange ).toHaveBeenNthCalledWith( 3, {
+			expect( props.onChange ).toHaveBeenNthCalledWith( 4, {
```
What is causing the `onChange` prop to be called one more time? Could it be related to the fact that [`user.type` also clicks on the element](https://testing-library.com/docs/user-event/utility#type)? If that's the case, we could try passing the `skipClick` option, or use `user.key` instead (since the input already has focus).

Same applies to other instances of this PR where the number of times that the `onChange` prop was called changed.

andrewhayward replied:
It's because of the additional `user.clear` call, which generates a `change` event of its own. Without it, `user.type` appends to the current value, rather than replacing it (which makes sense).

Did a bit of digging though, and `user.type` also accepts a selection range, which means the value of the field is totally replaced. This is closer to the original intent of directly firing a `change` event, so while `user.clear => user.type` is probably closer to most user interactions, `user.type` with a selection is probably preferable.

```javascript
await user.type( getWidthInput(), '0', {
  initialSelectionStart: 0,
  initialSelectionEnd: 1,
} );
```

ciampo replied:
Thank you for the digging, your proposed solution looks good :)

ciampo reviewed (commented):
Great job! Just left one comment, but once that's solved I believe this PR is an good place.

Could you also add an entry to the CHANGELOG? I'd say under the `Internal` section. Thank you!

Commit:
Replacing `user.clear` with a text selection range

Calling `user.clear` was necessary to prevent `user.type` from adding to an existing input value. However, by providing a selection range to `user.type` we can avoid that call entirely.

Commit:
Merge branch 'trunk' into internal/refactor-bordercontrol-unit-tests-fireevent-userevent

Commit:
Updating CHANGELOG.md

ciampo reviewed (approved):
LGTM 🚀 

Thank you for working on this one!
